File: server/src/service/EventBus.ts
import { EventEmitter } from "node:events";
import type { PaymentSession } from "../domain/session.js";

/**
 * Tiny pub/sub for session state changes. The SSE route subscribes per session
 * id and pushes the updated session to connected clients within ~1s of a change.
 */
export class EventBus {
  private emitter = new EventEmitter();

  constructor() {
    // A session can have several listeners (operator + display + history view).
    this.emitter.setMaxListeners(0);
  }

  publishSessionChange(session: PaymentSession): void {
    this.emitter.emit(`session:${session.id}`, session);
  }

  onSessionChange(id: string, listener: (session: PaymentSession) => void): () => void {
    const event = `session:${id}`;
    this.emitter.on(event, listener);
    return () => this.emitter.off(event, listener);
  }
}

File: android/app/src/main/java/cz/qrplatba/service/EventBus.kt
package cz.qrplatba.service

import cz.qrplatba.domain.PaymentSession
import java.util.concurrent.ConcurrentHashMap

/**
 * Tiny pub/sub for session state changes. The SSE route subscribes per session id
 * and pushes the updated session to connected clients on every change.
 */
class EventBus {
    private val listeners = ConcurrentHashMap<String, MutableSet<(PaymentSession) -> Unit>>()

    fun publishSessionChange(session: PaymentSession) {
        listeners[session.id]?.toList()?.forEach { it(session) }
    }

    /** Subscribe to changes for [id]; returns an unsubscribe function. */
    fun onSessionChange(id: String, listener: (PaymentSession) -> Unit): () -> Unit {
        val set = listeners.computeIfAbsent(id) { ConcurrentHashMap.newKeySet() }
        set.add(listener)
        return {
            set.remove(listener)
            if (set.isEmpty()) listeners.remove(id, set)
        }
    }
}
